[Cutlass] Add support for runtime param choices, starting with swizzle

mlazos opened the pull request:
This PR adds support for swizzle as a runtime parameter choice. Future runtime parameter choices can be added to the [get_runtime_arg_info](https://github.com/pytorch/pytorch/blob/2d40f9fb525350ac55486714e5620548f53b2958/torch/_inductor/codegen/cuda/cuda_template.py#L282) list method and then possible choices can be [looped over similarly to swizzle](https://github.com/pytorch/pytorch/blob/933f921b366f4c926ae1e653e81066e541f7d11b/torch/_inductor/codegen/cuda/gemm_template.py#L532). For precompile, we now filter choices by hash to only compile each distinct kernel source once. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #147224
* __->__ #147223
* #147222



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

mlazos committed:
[Cutlass] Add support for runtime param choices, starting with swizzle

[ghstack-poisoned]

mlazos commented:
re: testing, this is covered by the existing cutlass backend

mlazos committed:
Update on "[Cutlass] Add support for runtime param choices, starting with swizzle"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Add support for runtime param choices, starting with swizzle"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

Chillee commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -244,7 +246,13 @@ def def_kernel(
+        if runtime_arg_decls:
+            runtime_arg_decls += ", "
+
+        signature = f"int {self.kernel_name}({', '.join(arg_defs + size_args)}, {runtime_arg_decls}{self._EXTRA_CPP_ARGS})"
```
Probably worth asserting somewhere that these are all integers?

Chillee reviewed (approved):
Would be nice to have a test that it doesn't do extra recompilations, but nbd.

mlazos committed:
Update on "[Cutlass] Add support for runtime param choices, starting with swizzle"

This PR adds support for swizzle as a runtime parameter choice. Future runtime parameter choices can be added to the [get_runtime_arg_info](https://github.com/pytorch/pytorch/blob/2d40f9fb525350ac55486714e5620548f53b2958/torch/_inductor/codegen/cuda/cuda_template.py#L282) list method and then possible choices can be [looped over similarly to swizzle](https://github.com/pytorch/pytorch/blob/933f921b366f4c926ae1e653e81066e541f7d11b/torch/_inductor/codegen/cuda/gemm_template.py#L532). For precompile, we now filter choices by hash to only compile each distinct kernel source once. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison reviewed (approved):
Worth adding a run_and_get_code style test on the output code ?
